### PR TITLE
Retype the password in case login screen back, since sometimes the sc…

### PR DESCRIPTION
**Pull request by coolgw**
Retype the password in case login screen back

Retype the password in case login screen back, since sometimes the screen back to displaymanager, so we try to retype password, detail info please check poo#45236.
NOTE: I can not reproduce this issue on local machine, suspect the openqa performance issue?

- Related ticket: https://progress.opensuse.org/issues/45236
- Needles: na
- Verification run:
http://openqa.suse.de/tests/2864353#step/patch_sle/70
http://openqa.suse.de/tests/2864363#step/patch_sle/72

### CODE REVIEW

**Comment by lemon-suse** [comment]
Why not try several times, such as 3?

**Comment by coolgw** [comment]
try 3 times, update the code.

File: lib/opensusebasetest.pm [hunk 1]
```diff
@@ -569,6 +569,20 @@ sub wait_boot {
...
+                if (check_screen('displaymanager', 5) && check_var('DESKTOP', 'gnome')) {
+                    # In GNOME/gdm, we do not have to enter a username, but we have to select it
+                    if (is_tumbleweed) {
+                        send_key 'tab';
+                    }
+                    send_key 'ret';
+                    assert_screen 'displaymanager-password-prompt', no_wait => 1;
+                    type_password $password. "\n";
+                }
+                $count++;
```
Comment (lemon-suse) [inline lib/opensusebasetest.pm]: Missing break the while loop if get the expected result.

**Comment by okurz** [comment]
I am pretty sure the problem is caused by switching to the wrong tty because in newer versions of gdm there are two instances, once for login and the other instance which is the actually locked instance of the use. I f also created a ticket for that. The right approach is to switch to the right tty, not login again. Second, If the login fails but we type the correct password that's a product bug! Please handle as such. Third please find a solution which does not increase the complexity that much, e.g. less level of indentation

File: lib/opensusebasetest.pm (hunk 1 above)
Comment (coolgw) [reply lib/opensusebasetest.pm]: Need balance whether add break or not.
If not add break,  the function will try catch up gdm page three times and retry pass.
If add break, you have one try when encounter gdm page.
I prefer not add break.

**Comment by coolgw** [comment]
After some investigation, we suspect this is an issue and i have submitted an issue Bug 1127317 - desktop can not show after gdm window, and also add soft_fail flag into this workround script.
Since any failed case in patch_sle will impact our migration automation process, so i suggest using this workround firstly, then if the issue fixed then we disable this workround.

**Comment by coolgw** [comment]
@rwx788 could you help have a look? You can check verify result and this workround indeed start work now.

**Comment by coolgw** [comment]
@okurz  Could you have a look ? I plan merge this once this kind of issue impact our migration task check.

**Comment by okurz** [comment]
Sorry for not attending this in time (due to my longer absence). I see that you record a reference to https://bugzilla.suse.com/show_bug.cgi?id=1127317 . That is good already. However I strongly suggest to avoid the use of `check_screen` calls with non-zero timeouts. I hope you can find a way to use a multi-tag `assert_screen` with `match_has_tag` instead of `check_screen` with non-zero timeout to prevent introducing any timing dependant behaviour, to save test execution time as well as state more explicitly from the testers point of view what are the expected alternatives. For example:

```perl
assert_screen([qw(yast2_console-finished yast2_missing_package)]);
if (match_has_tag('yast2_missing_package')) {
    send_key 'alt-o';  # confirm package installation
    assert_screen 'yast2_console-finished';
}
```

The current way is very likely to introduce unpleasant, confusing surprises.

The tickets I failed to mention earlier which are about the original problem are the following:

* https://progress.opensuse.org/issues/48110
* https://progress.opensuse.org/issues/45710
* https://progress.opensuse.org/issues/44153

**Comment by SergioAtSUSE** [comment]
Reopen if needed: https://openqa.opensuse.org/tests/1013759#next_previous

**Pull request closed without merging**